hkmc2/shared/src/test/mlscript-compile/Term.mls

```diff
@@ -0,0 +1,192 @@
+
+abstract class DebrujinSplit : (Binder | DebrujinBranch | Accept | Reject)
+class Binder(body: DebrujinSplit) extends DebrujinSplit
+class DebrujinBranch(scrutinee: Int, ptrn, consequent: DebrujinSplit, alternative: DebrujinSplit) extends DebrujinSplit
+class Accept(outcome: Int) extends DebrujinSplit
+class Reject extends DebrujinSplit
+
+abstract class Pattern : (LitPattern | Var | ClassLike | Synonym | Tuple | Record)
+class LitPattern(lit: Literal) extends Pattern
+class Var(sym: Symbol) extends Pattern
+class ClassLike(sym: Symbol, trm: Term, parameters: globalThis.Array[Symbol] | null) extends Pattern
+class Synonym(symbol: Symbol, patternArguments: globalThis.Array[DebrujinSplit]) extends Pattern
+class Tuple(size: Int, inf: Bool) extends Pattern
+class Record(entities: globalThis.Array[(Str, Symbol)]) extends Pattern
+
+class Branch(scrutinee: Ref, ptrn: Pattern, continuation: Split)
+
+abstract class Split : (Cons | Let | Else | End)
+class Cons(head: Branch, tail: Split) extends Split
+class Let(sym: Symbol, term: Term, tail: Split) extends Split
+class Else(default: Term) extends Split
+class End extends Split
+
+
+object KeywordIf
+object KeywordWhile
+
+
+abstract class Statement : (Term | LetDecl | DefineVar)
+class LetDecl(sym: Symbol) extends Statement
+class DefineVar(sym: Symbol, rhs: Term) extends Statement
+// TODO: other statements
+
+
+abstract class Term : (Lit | Builtin | Ref | CSRef | App | Sel | DynSel | Tup | IfLike | Lam | Blk | New |
+  Region | RegRef | Assgn | Deref | SetRef | Ret | Throw | Try) extends Statement
+class Lit(lit: Literal) extends Term
+class Builtin(name: Str) extends Term
+class Ref(sym: Symbol) extends Term
+class CSRef(sym: Symbol, base: Str, file: Str | undefined) extends Term
+class App(lhs: Term, rhs: Term) extends Term
+class Sel(prefix: Term, nme: Str) extends Term
+class DynSel(prefix: Term, fld: Term, arrayIdx: Bool) extends Term
+class Tup(fields: globalThis.Array[Term]) extends Term
+class IfLike(kw: KeywordIf | KeywordWhile, desugared: Split) extends Term
+class Lam(params: globalThis.Array[Symbol], body: Term) extends Term
+class Blk(stats: globalThis.Array[Statement], res: Term) extends Term
+class New(cls: Term, args: globalThis.Array[Term]) extends Term
+class Region(name: Symbol, body: Term) extends Term
+class RegRef(reg: Term, value: Term) extends Term
+class Assgn(lhs: Term, rhs: Term) extends Term
+class Deref(ref: Term) extends Term
+class SetRef(ref: Term, value: Term) extends Term
+class Ret(result: Term) extends Term
+class Throw(result: Term) extends Term
+class Try(body: Term, finallyDo: Term) extends Term
+
+
+let builtinSymbols = new globalThis.Set()
+builtinSymbols.add("+")
+builtinSymbols.add("-")
+builtinSymbols.add("*")
+builtinSymbols.add("/")
+builtinSymbols.add("==")
+builtinSymbols.add("<")
+builtinSymbols.add(">")
+builtinSymbols.add(">=")
+builtinSymbols.add("<=")
+
+
+let names = new globalThis.Map()
+fun freshName(name) =
+  if names.has(name) == false do
```

Comment:
```suggestion
  if not(names.has(name)) do
```

Reply:
Please explain why you didn't apply the suggestion before resolving the conversation. I don't like when you do this.

https://github.com/hkust-taco/mlscript/pull/279/files#diff-070cb7224c53a3d478f0a055952d7646b2a74b9991eba9f085cf7d4f0c109eebR100

Reply:
Here, `not` is defined in `Predef.mls`. Since `Predef.mls` imported `Term.mls`, `Term.mls` importing `Predef.mls` will lead to cycle dependency (I think it was introduced in a recent commit). That's why I only used `is false` instead of `not`.

Reply:
There's no problem with such cyclic dependencies in JS actually. Please add `not` to `Predef` and use it.

And anyway, you should use the new `runtimeSymbol` approach to access the `Term` module. We want to move away from the ugly `globalThis.Predef` hack and instead use code like `Select(Value.Ref(State.runtimeSymbol), Tree.Ident("Term"))`. To do this you'll have to move the Term forwarding value from `Predef.mls` to `Runtime.mls`.

Reply:
Fixed.